Sources/NIO/Codec.swift

```diff
@@ -448,6 +453,9 @@ extension ByteToMessageHandler {
         var allowEmptyBuffer = decodeMode == .last
         while (self.state.isActive && self.removalState == .notBeingRemoved) || decodeMode == .last {
             let result = try self.withNextBuffer(allowEmptyBuffer: allowEmptyBuffer) { decoder, buffer in
+                if self.maximumBufferSize > 0, buffer.readableBytes >= self.maximumBufferSize {
+                    throw ByteToMessageDecoderError.payloadTooLarge
+                }
```

Comment:
could we move the check to after we try to decode? Just to give the decoder 'one last chance' because we loaded everything in memory already anyway so we might just as well give the decoder a try.

Just be sure you _don't_ use `defer`, this check must not run if we already threw an error

Reply:
@tomerd sure, but we load n+100 into memory _anyway_. So sure, the user intent is to never load more than `n` bytes into memory. But the current implementation will still load `n+100` into memory but it _won't_ give the decoder a last chance. I'm happy with saying that `n` isn't an absolute limit but best effort but we should give the decoder one last chance.

It's totally fine for this to be a best effort limit because we can't really implement anything else anyway.

Reply:
yes, you are absolutely correct from an “how much data is loaded to memory” pov. i am also thinking about the pov of the user wanting to only ever process n bytes and no more, for example if they have downstream logic that ties the expected max size to some preallocated space or simply tests for honoring max size

we could make this more complicated and introduce a behavior of type but i feel it would be an overshot. another alternative could be to truncate at max size if we buffered over it

other ideas/thoughts?

Reply:
@tomerd the user should never expect 'X and no more' because that is very rarely what we can do. If their parsing algorithm is so shit that they can't tolerate a few extra bytes, then I believe they should implement a check themselves using `guard buffer.readableBytes <= whatICanTolerate else { throw .... }`